lib/cn/csched_sp3.c

```diff
@@ -1797,43 +1797,45 @@ static void
 sp3_qos_check(struct sp3 *sp)
 {
     struct cn_tree *tree;
-    uint rootmin, rootmax;
     uint64_t rspill_dt_max;
+    int rootmin, rootmax;
     uint64_t sval;
 
     if (!sp->throttle_sensor_root)
         return;
 
     rootmin = sp->thresh.rspill_runlen_max;
-    rootmax = rootmin;
+    rootmax = 0;
     rspill_dt_max = 0;
     sval = 0;
 
     list_for_each_entry(tree, &sp->mon_tlist, ct_sched.sp3t.spt_tlink) {
         const struct kvs_rparams *rp = cn_tree_get_rp(tree);
-        uint marked, nk;
+        int split_cnt, nk;
 
         if (rp->cn_maint_disable)
             continue;
 
-        /* Discount kvsets currently being spilled (i.e., those that have
-         * been marked for spill by sp3_work()).
-         */
-        marked = (atomic_read_acq(&tree->ct_root->tn_busycnt) & 0xffffu) / 3;
-
         nk = cn_ns_kvsets(&tree->ct_root->tn_ns);
-        if (nk >= marked)
-            nk -= marked;
 
-        if (nk > rootmax) {
-            rspill_dt_max = max_t(uint64_t, rspill_dt_max, tree->ct_rspill_dt);
-            rootmax = nk;
+        /* Increase the throttle if a split is pending or in-progress as
+         * the root spill job count is either zero or approaching zero.
+         */
+        split_cnt = tree->ct_root->tn_split_cnt;
+        if (split_cnt > 0)
+            nk += (split_cnt / 2) + 1;
+
+        if (nk > rootmin) {
+            if (tree->ct_rspill_dt * (nk - rootmin) > rspill_dt_max * rootmax) {
+                rspill_dt_max = tree->ct_rspill_dt;
+                rootmax = nk - rootmin;
```

Comment:
seems like rootmax/rootmin should have the same type `nk`, which I think is uint32_t

Reply:
Yeah, I had changed them to signed because they could go negative, I can return them to uint.

Reply:
Well if they can go negative, then definitely don't make them unsigned! Change it to `int32_t` or something.

Reply:
They can't go negative now, so I changed them to uint32_t as you requested.